Adding spell-check command to SDK

roysagi opened the pull request:
Adding spell-check command to SDK.

Note: this also required adding "pyspellchecker" package to requirements.

The change in content_creator came from the precommit hooks

fixes: https://github.com/demisto/etc/issues/23706

ronykoz commented:
@guykeller let me know when you finish reviewing this

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            int. 0 if no problematic words found, 1 otherwise.
+        """
+        unknown_words = set([])
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
```
This line simply making these words known to the spell checker, right? 
If so let's add an explaining comment to it

roysagi replied:
done

ronykoz replied:
maybe this should be done in the init

roysagi replied:
moved this to the attributes.

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+
+            else:
+                if isinstance(value, dict):
+                    if key != SCRIPT_ARGS:
```
Could you please explain this condition?

ronykoz replied:
scriptarguments is something we should not explore, @roysagi lets add a comment about this

roysagi replied:
its the field name for argument for commands in playbooks. this line is ok, I'll just add an explanation.

guykeller commented on demisto_sdk/common/constants.py:
```diff
@@ -690,3 +690,22 @@ class PB_Status:
+    "IOCs",
+    "Emails",
+    "Google"
+]
```
Let's discuss known words @roysagi @ronykoz

ronykoz replied:
it should be a file with words like Demisto, Dbotscore,... and we should have an option to provide our own file using a param

guykeller commented:
@roysagi Let's further review the logic and understand how this is working at the moment

roysagi commented on demisto_sdk/common/constants.py:
```diff
@@ -690,3 +690,22 @@ class PB_Status:
+    "comment"
+]
+
+SCRIPT_ARGS = 'scriptarguments'
```
done

ronykoz replied:
i think it should be hardcoded in the class itself perhaps

ronykoz commented on demisto_sdk/common/constants.py:
```diff
@@ -690,3 +690,22 @@ class PB_Status:
     'url': {'URL.Data'}
 }
+
+DISPLAYABLE_LINES = [
```
add a comment here or think of a better name please

ronykoz commented on demisto_sdk/core.py:
```diff
@@ -158,3 +162,15 @@ def create_content_artifacts(self, artifact_path, preserve_bundles):
+            path (str): The path to the checked file.
+
+        Returns:
+          int. 0 if no problematic words found, 1 otherwise.
```
use boolean

ronykoz commented on demisto_sdk/core.py:
```diff
@@ -72,6 +74,8 @@ def parse_args(self):
             elif args.command == 'create':
                 self.create_content_artifacts(args.artifacts_path, args.preserve_bundles)
+            elif args.command == 'spell-check':
+                self.spell_check(path=args.path)
```
we are not doing anything with it now, please add a TODO comment to fix this

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+
+
+class SpellCheck:
+    """
```
add an explanation for the class itself

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+import yaml
+
+from spellchecker import SpellChecker
```
move this to line 2, keep the other ones separated by a line

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            with open(self.path, 'r') as md_file:
+                md_data = md_file.readlines()
+
+            self.check_md_file(md_data=md_data, unknown_words=unknown_words)
```
we should have a space both before and after

roysagi replied:
done

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+        """Runs spell-check on the given file.
+
+        Returns:
+            int. 0 if no problematic words found, 1 otherwise.
```
it's python, use boolean

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            unknown_words (set): A set of found unknown words
+
+        Returns:
+            None.
```
remove

roysagi replied:
done

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            None.
+        """
+        for key, value in yml_info.items():
+            if key in DISPLAYABLE_LINES:
```
yes, other fields could be field names etc, which are not relevant

roysagi replied:
made the name more descriptive and added further explanation.

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            else:
+                if isinstance(value, dict):
+                    if key != SCRIPT_ARGS:
+                        self.check_yaml(self.spellchecker, value, unknown_words)
```
add blank line after

roysagi replied:
done

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+        from argparse import ArgumentDefaultsHelpFormatter
+        description = """Run spell check on a given yml/md file. """
+        parser = subparsers.add_parser('spell-check', help=description, formatter_class=ArgumentDefaultsHelpFormatter)
+        parser.add_argument("-p", "--path", help="Specify path of yml/md file", required=True)
```
add empty line at the end of the file

roysagi replied:
done

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+
+    @staticmethod
+    def add_sub_parser(subparsers):
+        from argparse import ArgumentDefaultsHelpFormatter
```
why is this here?

roysagi replied:
it is the way every add_sub_parser function is built in every class in the sdk.

ronykoz replied:
@liorblob is there a good reason for this?

liorblob replied:
it's not in every class, it was just in the linter... no actual reason for it to be in every class.

roysagi replied:
ok then... I'll take it out

roysagi commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            return 1
+
+        if unknown_words:
+            print_error(u"Found the problematic words in {}:\n{}".format(self.path, '\n'.join(unknown_words)))
```
done

roysagi commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+            print_error(u"Found the problematic words in {}:\n{}".format(self.path, '\n'.join(unknown_words)))
+            return 1
+
+        print_color("No problematic words found in {}".format(self.path), LOG_COLORS.GREEN)
```
done

roysagi commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,95 @@
+        print_color("No problematic words found in {}".format(self.path), LOG_COLORS.GREEN)
+        return 0
+
+    def check_md_file(self, md_data, unknown_words):
```
changed

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
 
 class SpellCheck:
-    """
+    """Preform a spell check on the given .yml or .md file.
```
Perform*

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
         """
-        unknown_words = set([])
-        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
+        # adding known words file if given
```
Let's add that known words means they don't count as misspelled words

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
 
-    def __init__(self, path: str):
+    def __init__(self, path: str, known_words: str):
         self.path = path
```
Let's change this name to checked_file_path

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
+            self.spellchecker.word_frequency.load_text_file(known_words_file)
 
+        # adding the KNOWN_WORDS to the spellchecker recognized words.
+        self.spellchecker.word_frequency.load_words(KNOWN_WORDS)
```
Let's move this logic to an add_known_words helper function

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
         if self.path.endswith('.md'):
             with open(self.path, 'r') as md_file:
-                md_data = md_file.readlines()
+                md_file_lines = md_file.readlines()
```
Let's move these lines into check_md_file

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
-            self.check_md_file(md_data=md_data, unknown_words=unknown_words)
         elif self.path.endswith('.yml'):
             with open(self.path, 'r') as yaml_file:
                 yml_info = yaml.safe_load(yaml_file)
```
Let's move these lines into check_yaml

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
+        # The following code separates the spell check to these cases.
         for key, value in yml_info.items():
-            if key in DISPLAYABLE_LINES:
+            # case 1: a String.
```
Let's change to case 1:string value of a user visible field or something like that

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
                 for word in value.split():
+
+                    # Drop commas, full-stops and brackets.
+                    if word.endswith(',') or word.endswith('.') or word.endswith(')'):
```
What if it ends with ":" or "-"?

roysagi replied:
:  I agree... but why -?

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -1,95 +1,129 @@
                 if isinstance(value, dict):
-                    if key != SCRIPT_ARGS:
-                        self.check_yaml(self.spellchecker, value, unknown_words)
+                    # ignore command arguments in playbooks - no need to check these.
```
Let's add reference to "scriptarguments" here

roysagi replied:
it's kind of long - I dont think we want to do it
scriptarguments:
      message-id:
        complex:
          root: Gmail
          filters:
          - - operator: isNotEmpty
              left:
                value:
                  simple: Gmail.Attachments
                iscontext: true
          transformers:
          - operator: GetFirstObject
          - operator: getField
            args:
              field:
                value:
                  simple: ID

guykeller replied:
I meant we should add a small reference to what "scriptarguments" is

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,126 @@
+class SpellCheck:
+    """Perform a spell check on the given .yml or .md file.
+        Attributes:
+            path (str): The path to the current file being checked.
```
```suggestion
            checked_file_path (str): The path to the current file being checked.
```

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,126 @@
+
+        elif self.checked_file_path.endswith('.yml'):
+            with open(self.checked_file_path, 'r') as yaml_file:
+                yml_info = yaml.safe_load(yaml_file)
```
Let's add this code to the check_yaml method

roysagi replied:
I don't think we want to do that. The check_yml function is recursive and works by getting a dictionary.
This is simply the initialisation of the first dictionary, putting it there will add an unnecessary step in my opinion.

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,126 @@
+                        "Only .yml or .md files are supported.".format(self.checked_file_path))
+            return False
+
+        if self.unknown_words:
```
```suggestion
        if len(self.unknown_words) > 0:
```

roysagi replied:
done

guykeller commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,126 @@
+                # case 3: a list of dictionaries
+                elif isinstance(value, list):
+                    for sub_dict in value:
+                        if isinstance(sub_dict, dict):
```
Are we sure all of the sub_dicts will be a dict and none will be a list?

roysagi replied:
I parsed both an integration and a playbook.
In an integration, I didn't find '[[' - so no nested lists as far as I can tell.
In a playbook the only '[[' I found were for operators and thats under "scriptarguments" which we ignore - like so:
'**scriptarguments**': {'format': {}, 'thread-id': {'complex': {'root': 'Gmail', 'filters': **[[**{'operator': 'isEqualString', 'left': {'value': {'simple': 'Gmail.Mailbox'}, 'iscontext': True}, 'right': {'value': {'simple': 'admin@demistodev.com'}}}]]...

roysagi commented:
Added nltk package (https://www.nltk.org/).
Specifically I added the 'words' sub-package which is downloaded to the user's computer upon running spellcheck (so the first run is a little slower but not by too much).

There is a bigger dictionary ('brown') but it's about 8G so it seems a bit excessive - but we can add this to first installation and then we can use it.

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,148 @@
+    """Perform a spell check on the given .yml or .md file.
+        Attributes:
+            checked_file_path (str): The path to the current file being checked.
+            known_words (str): The path to a file containing known words.
```
known_words_file_path

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,148 @@
+        self.checked_file_path = checked_file_path
+        self.spellchecker = SpellChecker()
+        self.unknown_words = set([])  # type:Set
+        self.known_words = known_words
```
same_here

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,148 @@
+        except AttributeError:
+            pass
+        else:
+            ssl._create_default_https_context = _create_unverified_https_context
```
what is happening here

roysagi replied:
added a description for this move as well as the download

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,148 @@
+            ssl._create_default_https_context = _create_unverified_https_context
+
+        # downloading words set from nltk.
+        nltk.download('words')
```
why do we need this if we have the corpus of words?

roysagi replied:
see above

ronykoz commented on demisto_sdk/common/constants.py:
```diff
@@ -695,3 +695,13 @@ class PB_Status:
     'url': {'URL.Data'}
 }
+
+KNOWN_WORDS = [
```
we should have this as a file and read it, add words like Demisto, DBotScore please

roysagi replied:
moved out of constants to a different file

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,148 @@
+        # adding known words file if given - these words will not count as misspelled
+        if self.known_words:
+            known_words_file = open(self.known_words, 'r')
+            self.spellchecker.word_frequency.load_text_file(known_words_file)
```
use with statement please
you are not closing the file and it will close it automatically when using this

ronykoz commented on demisto_sdk/dev_tools/spell_checker.py:
```diff
@@ -0,0 +1,150 @@
+        description = """Run spell check on a given yml/md file. """
+        parser = subparsers.add_parser('spell-check', help=description, formatter_class=ArgumentDefaultsHelpFormatter)
+        parser.add_argument("-p", "--path", help="Specify path of yml/md file", required=True)
+        parser.add_argument("--known_words", help="A file path to a txt file with known words")
```
```suggestion
        parser.add_argument("--known_words", help="A file path to a txt file with known words to exclude in the check")
```

roysagi commented:
Will add Unit tests before merging.

ronykoz commented on requirements.txt:
```diff
@@ -1,10 +1,13 @@
 bandit>=1.6.2
 bs4>=0.0.1
+bs4>=0.0.1
```
you have duplicates

ronykoz commented on docs/spell_check_command.md:
```diff
@@ -0,0 +1,24 @@
+### Spell-check
```
@dantavori apply your style changes here too please

ronykoz commented:
@yaakovi maybe we should add this one?

Itay4 commented:
@ronykoz is this still relevant ?
